core/dbt/cli/main.py

```diff
@@ -25,6 +28,52 @@ def cli_runner():
     cli()
 
 
+class dbtUsageException(Exception):
+    pass
+
+
+# Programmatic invocation
+class dbtRunner:
+    def __init__(self, project: Project = None, profile: Profile = None):
+        self.project = project
+        self.profile = profile
+
+    def invoke(self, args):
+        dbt_ctx = cli.make_context(cli.name, args)
+        dbt_ctx.obj = {}
+        dbt_ctx.obj["project"] = self.project
+        dbt_ctx.obj["profile"] = self.profile
+
+        try:
+            return cli.invoke(dbt_ctx)
+        except (click.NoSuchOption, click.UsageError) as e:
+            raise dbtUsageException(e.message)
+
+
+# TODO: refactor - consider decorator, or post-init on a dbtContext
+def _initialize_context(ctx):
+    flags = Flags()
+
+    # Tracking
+    initialize_from_flags(flags.ANONYMOUS_USAGE_STATS, flags.PROFILES_DIR)
+    ctx.with_resource(track_run(run_command=flags.WHICH))
+
+    ctx.obj = ctx.obj or {}
+
+    # Profile
```

Comment:
I think we talked about having some sort of mapping defining which command would need what object(Deps doesn't need profile for example). Should we define those in the actual command and use it here so that we don't always load all objects?

Also thinking how can we expose that to server so that they don't need to fetch things that is not needed during initialize a task.

I think all I am talking about here can be a follow up thing after the current change.

Reply:
This actually resolved itself quite nicely via [`requires.profile`, `requires.project`](https://github.com/dbt-labs/dbt-core/pull/6387/files#diff-43832707bfec8cdaa49c13f004d9b3296585827f7e68ef07fba7ad1d5e6eb7c6R234-R236) decorators once we pivoted to initializing flags / context objects at the subcommand level.